### PR TITLE
admin can publish a 'draft' round, and choose whether to open it for proposals or contributions

### PR DESCRIPTION
**Pull request by euglazer**
rounds are either a 'draft' or 'published'. 

a published round can be in one of three 'modes'. (1) proposal mode, (2) contribution mode, or (3) closed mode.

email notifications depend on these 'modes' and these 'modes' depend on whether the round is a 'draft' or 'published'. 

so, i've added a '#mode' method and 'published' attribute to the Round model.

I've also added a '#publish!' method, which updates the 'published' attribute to 'true'. later, it will send a notification email to all relevant users as described by @derekrazo in #71 

@derekrazo @alannallama @ahdinosaur @robguthrie 
i'd like to double check this logic with y'all before I move on with the email notification stuff.

lmk what ya think, thx!

### CODE REVIEW

**Comment by ahdinosaur** [comment]
looks good to me. :+1: 

our previous logic for this was instead of having a 'published' field, whether a round was a draft or not was based on whether or not there was a 'starts_at' *and* 'ends_at' field. took me a sec, but now i see why you did it this way, sweet. :strawberry: 

an open question is how we handle going from `draft` mode immediately to `contribution` mode, as it is it would only work if you specifically set the `starts_at` field to be before the current time. food for thought. :bouquet:

**Comment by euglazer** [comment]
thx mikey.

ya good point. i'm thinking if the admin wanted to immediately enter ```contribution``` mode after publishing, i could make a method / controller action that sets ```starts_at``` to the current time.

File: spec/factories/rounds.rb [hunk 1]
```diff
@@ -19,6 +28,7 @@
     members_can_propose_buckets true
```
Comment (euglazer) [inline spec/factories/rounds.rb]: if a round is open for contributions, shouldn't `members_can_propose_buckets` be false?

File: spec/factories/rounds.rb [hunk 2]
```diff
@@ -27,5 +37,6 @@
...
```
Comment (euglazer) [inline spec/factories/rounds.rb]: similarly, if a round is closed, shouldn't `members_can_propose_buckets` be false?

if this is the case, then can't `members_can_propose_buckets` be a computed value based on whether the round is in 'proposal mode' or not?

Comment (euglazer) [inline spec/factories/rounds.rb]: `members_can_propose_buckets` doesn't seem to be used at all on the back-end

Comment (ahdinosaur) [inline spec/factories/rounds.rb]: `members_can_propose_buckets` is whether or not the round *should have* a `proposal` mode, not whether or not it *is in* a `proposal` mode.

Comment (ahdinosaur) [inline spec/factories/rounds.rb]: but yeah, we never got around to actually implementing it properly. :stuck_out_tongue_closed_eyes:

**Comment by euglazer** [comment]
publish route complete with specs.

##### brief summary of my changes:
added a `put '/rounds/:round_id/publish'` route requiring a combination of the following parameters:
```
{
  "round" : {
    "skip_proposals" : <boolean>,
    "starts_at" : <datetime>,
    "ends_at" : <datetime>
  }
}
```
when you publish a round, you decide whether you'd like to open the round for proposals, or skip proposals and immediately open the round for contributions. 

if you select `skip_proposals: false` and open the round for proposals, the round's `published` attribute is set to `true`, and its `starts_at` and `ends_at` times are set to those specified in the input parameters. notification emails are then sent out to each member of the round. that email contains a link to the 'buckets' page.

if you select `skip_proposals: true` and immediately open the round for contributions, the round's `published` attribute is set to `true`, its `starts_at` is set to the time right now, and its `ends_at` is set to the time specified in the input parameters. an email is then sent out to each member with allocations in that round. that email links them to the 'buckets' page, and tells them how much money they have to spend.

this route is only accessible to an admin.

@ahdinosaur @alannallama @derekrazo @robguthrie 
mergeable?

**Comment by ahdinosaur** [comment]
at a high-level looks good to me. :+1:

**Comment by ahdinosaur** [comment]
@euglazer 

1. friendly reminder to push the changes we made during our pair sesh today, want to make sure those aren't lost as they're necessary for the whole thing to work.
2. would you be able to add info to the README about how to run the mailer locally? it seems to be a bit more involved now with redis, sidekiq and such, so would be good to have documented.

**Comment by robguthrie** [comment]
Are you able to use postgres rather than having to run redis?

**Comment by euglazer** [comment]
@ahdinosaur 
documentation updated

@robguthrie 
i don't know -- if we don't have to use redis that's sweet. all the tutorials i looked at were using sidekiq with redis. would be down to simplify, though I don't immediately know how. will give it a go today

**Comment by robguthrie** [comment]
Ok cool. Well having to setup/run/pay for a redis server just to send
emails is a bit overkill I think. I think you're writing nice ActiveJob
based jobs, in which case changing the backend is really simple. I would
recommend you use something which can run off the same database that you're
already using (I've had a reasonably good time with DelayedJob in Loomio).
I'm referencing http://edgeguides.rubyonrails.org/active_job_basics.html

**Comment by euglazer** [comment]
sweet - will prioritize it! thx rob

**Comment by robguthrie** [comment]
If you wanted to .. you could just use Acitve Job Inline for a short period
of time, and leave all the server fiddling work for another day.

**Comment by euglazer** [comment]
@robguthrie 

delayed job is working so well so far! thx rob!

**Comment by euglazer** [comment]
FINALLY THE DAMN THING BUILDS

**Comment by robguthrie** [comment]
This looks great Eugene!

**Comment by robguthrie** [comment]
booo ya!

**Comment by derekrazo** [comment]
Awesome work team :)

**Pull request merged**